### PR TITLE
External Resource Widget

### PR DESCRIPTION
**Pull request by 3nids**
This creates a new editor widget: External Resource Widget

Following PR #2547, this improves the current File Name Widget

The decision has been made to create a new editor to:
- be able to merge with photo widget
- avoid duplicating code for legacy support
- properly access the widget through the API and the custom widgets

Users will be invited to use the new editor in the config dialog, and old widgets will be removed in QGIS 3 (with automatic migration to new editor).

To be done:
- [x] merge photo widget
- [ ] display warnings in configs of old widget
- [ ] migration to new editor
- [ ] SIP binding and custom widget
- [x] use QgsGroupBox instead of QGroupBox in config dialog

### CODE REVIEW

**Comment by 3nids** [comment]
@medspx can you have a try?

**Comment by nyalldawson** [comment]
Just want to register my interest in this work - I'll try to post my thoughts asap, but please don't merge in the meantime

**Comment by nyalldawson** [comment]
@3nids Ok, so what I've been thinking for a while is that we need a generic QGIS widget for "filename" line edits. Currently there's a huge mix of different implementations of this throughout QGIS, each with their own unique behaviour and UI. 

Possible QgsExternalResourceWidget could be extended to handle this, or (probably better) a new QgsFileLineEditWidget could be created which would be utilised by QgsExternalResourceWidget . 

Requirements would include:
- support for both input and output file name modes - this would change the behaviour of the browse file dialog accordingly
- support for multiple file open modes (eg for the add vector layer dialog)
- allow drag and drop of files onto widget to set the filename
- have some handling of default paths. Eg, if the user only enters a filename in the widget, it needs logic to expand to this a sensible full path
- support for setting the valid file formats shown in the browse file dialogs

What do you think? Does this sound sensible to you?

**Comment by 3nids** [comment]
Yep, this sounds like a good idea.
I'll do it, but it may need more time.

**Comment by 3nids** [comment]
it will be nice to allow multiple file support for fields too.

**Comment by medspx** [comment]
Yes, file filters will also be a good feature for ~~File Name~~ QgsExternalResource widget (I thought to add it after the merge of my PR).
I have also another feature request (but I am not sure that it can be handled by QgsExternalResource): some users also want to store URL into a field and be able to click on links to open those URLs in web browser.

Well, it is very nice to see some changes on these widgets, it will made QGIS better...

**Comment by 3nids** [comment]
I think your changes could be integrated too.
I am thinking that the document viewer should not be only for pictures, but fore PDF, web pages, etc.
Now, we can have the QgsFilePickerWidget that only deals with files, then QgsExternalResource (the editor) can integrate the file picker, the url line edit and the document viewer.

**Comment by nyalldawson** [comment]
> Now, we can have the QgsFilePickerWidget that only deals with files, then QgsExternalResource (the editor) can integrate the file picker, the url line edit and the document viewer.

That was my thoughts too.

**Comment by m-kuhn** [comment]
Will this be part of 2.14?

**Comment by 3nids** [comment]
I'll try. Freeze is on 15th, right?

**Comment by 3nids** [comment]
OK, I just completed this PR.

Right now QgsFilePicker can only pick single file or folder. I will either improve it for QGIS 3 or @nyalldawson if you want to do it, go ahead.
QgsExternalRessourceWidget combines the file picker with an optional document viewer (photo or web). It could extended to display PDF or gallery for multiple images.

I still have a few things to polish (sip, custom widget, warning message for deprecated widgets), but this can be tested and/or reviewed. I am totally open to better phrasing and naming of classes/methods.

File: src/gui/qgsfilepickerwidget.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  else if ( !mFilePath.isEmpty() )
+  {
+    oldPath = relativePath( mFilePath, false );
+  }
+
+  // If there is no valid value, find a default path to use
+  QUrl theUrl = QUrl::fromUserInput( oldPath );
+  if ( !theUrl.isValid() )
+  {
+    oldPath = settings.value( "/UI/lastExternalResourceWidgetDir", QDir::cleanPath( QgsProject::instance()->fileInfo().absolutePath() ) ).toString();
```
Comment (nyalldawson) [inline src/gui/qgsfilepickerwidget.cpp]: What's the default path here if a project hasn't been saved? It should default to something sensible in this case, eg the users' documents folder (and definitely not the annoying default qgis bin folder!)

**Comment by nyalldawson** [comment]
@3nids some comments regarding the QgsFilePicker implementation:

- in general looks very nice, exactly what I had in mind! Have you given any thought to a "destination" file/directory mode, rather than just the current "source" mode. Eg this would change the choose dialogs to use getSaveFileName instead. This would be handy for things like the raster calculator destination file name widget.
- would be good to see some unit tests of QgsFilePickerWidget::relativePath and QgsFilePickerWidget::toUrl . There's likely to be a fair bit of tweaking required to these functions to handle special cases and it would be nice to avoid any regressions
- There should be a way to set the title of the dialog, and the file filters for the dialog too

File: src/gui/qgsfilepickerwidget.cpp (hunk 1 above)
Comment (3nids) [inline src/gui/qgsfilepickerwidget.cpp]: indeed, it's now updated

**Comment by 3nids** [comment]
manually merged.

**Pull request closed without merging**